File: frontend/src/lib/review.ts
/**
 * Client-side API helpers for the /review endpoints (Phase 3).
 */
import { api, API_URL } from "@/lib/api";

// ── Types ────────────────────────────────────────────────────────────────────
export interface SiteIssues {
  below_award: number;
  no_proposed_rate: number;
  unknown_level: number;
}

export interface SiteSummary {
  site: string;
  staff: number;
  payroll_current: number;
  payroll_proposed: number;
  issues: SiteIssues;
  approval_status: string;
}

export interface CheckResult {
  status: "ok" | "warn" | "fail" | "suppressed";
  label: string;
  detail: string;
  recommendation: string;
}

export interface SuppressionInfo {
  id: number;
  check_label: string;
  reason: string | null;
  suppressed_by_name: string;
  suppressed_at: string; // ISO datetime string
  undone_by_name: string | null;
  undone_at: string | null;
}

export interface EmployeeCompliance {
  checks: CheckResult[];
  overall: "ok" | "warn" | "fail";
  award_minimum: number | null;
  next_level: string | null;
  band_min: number | null;
  band_max: number | null;
  suppressions: SuppressionInfo[];
}

export interface EmployeeWithCompliance {
  id: number;
  cycle_id: number;
  emp_num: string;
  first_name: string;
  last_name: string;
  email: string | null;
  age: number | null;
  site: string;
  department: string | null;
  category: string | null;
  hours_per_week: number | null;
  fy25_award: string | null;
  current_rate: number | null;
  fy26_award: string | null;
  proposed_award: string | null;  // accepted next-level suggestion
  pp_level: string | null;
  change_type: string | null;
  change_input: number | null;
  proposed_rate: number | null;
  letter_type: string | null;
  notes: string | null;
  is_departed: boolean;

  // Historical compliance snapshot (FY25→FY26, read-only, from Excel)
  hist_award_level_changed: boolean | null;
  hist_rate_changed: boolean | null;
  hist_above_award_rate: boolean | null;
  hist_above_pp_rate: boolean | null;
  hist_above_pp_max: boolean | null;

  compliance: EmployeeCompliance;
}

export interface EmployeePatch {
  // proposed_rate is calculated server-side — send change_type + change_input instead
  change_type?: string | null;
  change_input?: number | null;
  // Accepted next-level suggestion (empty string = clear)
  proposed_award?: string | null;
  // Other editable fields
  pp_level?: string | null;
  letter_type?: string | null;
  notes?: string | null;
}

export interface BulkSuggestResponse {
  updated: number;
  skipped: number;
}

export interface SubmitSiteResponse {
  site: string;
  status: "pending";
  issues_count: number;
}

// ── API calls ────────────────────────────────────────────────────────────────
export async function listSites(cycleId: number): Promise<SiteSummary[]> {
  return api<SiteSummary[]>(`/api/v1/cycles/${cycleId}/sites`);
}

export async function getSiteEmployees(
  cycleId: number,
  site: string,
): Promise<EmployeeWithCompliance[]> {
  return api<EmployeeWithCompliance[]>(
    `/api/v1/cycles/${cycleId}/sites/${encodeURIComponent(site)}/employees`,
  );
}

export async function patchEmployee(
  empId: number,
  patch: EmployeePatch,
): Promise<EmployeeWithCompliance> {
  return api<EmployeeWithCompliance>(`/api/v1/employees/${empId}`, {
    method: "PATCH",
    body: patch,
  });
}

export async function bulkSuggest(
  cycleId: number,
  site?: string,
): Promise<BulkSuggestResponse> {
  return api<BulkSuggestResponse>(`/api/v1/cycles/${cycleId}/bulk-suggest`, {
    method: "POST",
    body: { site: site ?? null },
  });
}

export async function submitSite(
  cycleId: number,
  site: string,
): Promise<SubmitSiteResponse> {
  return api<SubmitSiteResponse>(
    `/api/v1/cycles/${cycleId}/sites/${encodeURIComponent(site)}/submit`,
    { method: "POST" },
  );
}

export interface BulkAssignLettersResponse {
  updated: number;
  skipped: number;
}

export async function bulkAssignLetters(
  cycleId: number,
  site: string,
): Promise<BulkAssignLettersResponse> {
  return api<BulkAssignLettersResponse>(
    `/api/v1/cycles/${cycleId}/sites/${encodeURIComponent(site)}/assign-letters`,
    { method: "POST" },
  );
}

export async function suppressCheck(
  empId: number,
  checkLabel: string,
  reason?: string,
): Promise<EmployeeWithCompliance> {
  return api<EmployeeWithCompliance>(`/api/v1/employees/${empId}/suppress-check`, {
    method: "POST",
    body: { check_label: checkLabel, reason: reason ?? null },
  });
}

export async function unsuppressCheck(
  empId: number,
  checkLabel: string,
): Promise<EmployeeWithCompliance> {
  return api<EmployeeWithCompliance>(
    `/api/v1/employees/${empId}/suppress-check/${encodeURIComponent(checkLabel)}`,
    { method: "DELETE" },
  );
}

// ── Draft letter downloads ────────────────────────────────────────────────────
// Use programmatic fetch (credentials: "include") so the session cookie is sent
// to the FastAPI backend, then trigger a synthetic <a> click to save the blob.

async function _triggerBlobDownload(url: string, fallbackName: string): Promise<void> {
  const res = await fetch(url, { credentials: "include" });
  if (!res.ok) {
    const body = await res.json().catch(() => res.text());
    const detail =
      typeof body === "object" && body !== null && "detail" in body
        ? String((body as { detail: unknown }).detail)
        : typeof body === "string"
          ? body
          : `HTTP ${res.status}`;
    throw new Error(detail);
  }
  const blob = await res.blob();
  const disposition = res.headers.get("Content-Disposition") ?? "";
  const fnMatch = disposition.match(/filename="([^"]+)"/);
  const filename = fnMatch ? fnMatch[1] : fallbackName;

  const objectUrl = URL.createObjectURL(blob);
  const a = document.createElement("a");
  a.href = objectUrl;
  a.download = filename;
  document.body.appendChild(a);
  a.click();
  document.body.removeChild(a);
  URL.revokeObjectURL(objectUrl);
}

/** Download a single employee's DRAFT pay-letter PDF. */
export async function downloadDraftLetter(empId: number): Promise<void> {
  await _triggerBlobDownload(
    `${API_URL}/api/v1/employees/${empId}/draft-letter`,
    `draft-letter-${empId}.pdf`,
  );
}

/** Download all ready DRAFT pay-letter PDFs as a ZIP. */
export async function downloadDraftLettersZip(cycleId: number, site: string): Promise<void> {
  await _triggerBlobDownload(
    `${API_URL}/api/v1/cycles/${cycleId}/sites/${encodeURIComponent(site)}/draft-letters.zip`,
    `draft-letters-${site}.zip`,
  );
}

File: backend/app/schemas/review.py
"""Schemas for Phase 3 & 4 — review workflow + approvals endpoints."""
from __future__ import annotations

from datetime import datetime
from typing import Literal, Annotated

from pydantic import BaseModel, ConfigDict, Field


# ─────────────────────────────────────────────────────────────────────────────
#  Site summary (for /cycles/{id}/sites)
# ─────────────────────────────────────────────────────────────────────────────
class SiteIssues(BaseModel):
    below_award: int = 0
    no_proposed_rate: int = 0
    unknown_level: int = 0


# ─────────────────────────────────────────────────────────────────────────────
#  Per-check result (mirrors compliance.CheckResult dataclass)
# ─────────────────────────────────────────────────────────────────────────────
class CheckResult(BaseModel):
    status: Literal["ok", "warn", "fail", "suppressed"]
    label: str
    detail: str
    recommendation: str = ""


class SuppressionInfo(BaseModel):
    """Who suppressed a check and when."""
    id: int
    check_label: str
    reason: str | None
    suppressed_by_name: str
    suppressed_at: datetime
    # Undo info (only present after undo)
    undone_by_name: str | None = None
    undone_at: datetime | None = None


class SiteSummary(BaseModel):
    site: str
    staff: int
    payroll_current: float   # hours * current_rate * 52
    payroll_proposed: float  # hours * proposed_rate * 52 (0 where no proposed rate)
    issues: SiteIssues
    approval_status: str     # mirrors Approval.status


# ─────────────────────────────────────────────────────────────────────────────
#  Employee with compliance flags (for /cycles/{id}/sites/{site}/employees)
# ─────────────────────────────────────────────────────────────────────────────
class EmployeeCompliance(BaseModel):
    checks: list[CheckResult] = Field(default_factory=list)
    overall: Literal["ok", "warn", "fail"] = "ok"
    award_minimum: float | None = None
    next_level: str | None = None      # suggested reclassification target
    band_min: float | None = None
    band_max: float | None = None
    suppressions: list[SuppressionInfo] = Field(default_factory=list)


class EmployeeWithCompliance(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    cycle_id: int
    emp_num: str
    first_name: str
    last_name: str
    preferred_name: str | None = None
    email: str | None
    age: int | None
    site: str
    department: str | None
    category: str | None
    job_classification: str | None = None
    rate_type: str | None = None
    hours_per_pay_period: float | None = None
    hours_per_week: float | None
    current_award: str | None
    current_rate: float | None
    proposed_award: str | None    # accepted next-level suggestion
    pp_level: str | None
    change_type: str | None
    change_input: float | None
    proposed_rate: float | None
    letter_type: str | None
    notes: str | None
    is_departed: bool

    compliance: EmployeeCompliance = Field(default_factory=EmployeeCompliance)


# ─────────────────────────────────────────────────────────────────────────────
#  Employee PATCH
# ─────────────────────────────────────────────────────────────────────────────
class EmployeePatchRequest(BaseModel):
    # Review workflow — proposed_rate is calculated server-side from these two
    change_type: str | None = None
    change_input: float | None = None
    # Accepted next-level suggestion (empty string = clear)
    proposed_award: str | None = None
    # Other editable fields
    pp_level: str | None = None
    letter_type: str | None = None
    notes: str | None = None


# ─────────────────────────────────────────────────────────────────────────────
#  Bulk suggest
# ─────────────────────────────────────────────────────────────────────────────
class SuppressCheckRequest(BaseModel):
    check_label: str
    reason: str | None = None


class BulkSuggestRequest(BaseModel):
    site: str | None = None  # limit to one site; None = entire cycle


class BulkSuggestResponse(BaseModel):
    updated: int
    skipped: int  # employees already with a proposed rate (not overwritten)


class BulkAssignLettersResponse(BaseModel):
    updated: int   # employees whose letter_type was set / changed
    skipped: int   # employees with no proposed rate (can't infer)


# ─────────────────────────────────────────────────────────────────────────────
#  Site submission / approval
# ─────────────────────────────────────────────────────────────────────────────
class SubmitSiteRequest(BaseModel):
    site: str


class SubmitSiteResponse(BaseModel):
    site: str
    status: Literal["pending"]
    issues_count: int  # hard compliance errors remaining


class ApprovalDecisionRequest(BaseModel):
    decision: Literal["approve", "request_changes"]
    comment: str | None = None


class ApprovalDecisionResponse(BaseModel):
    site: str
    status: Literal["approved", "changes_requested"]


# ─────────────────────────────────────────────────────────────────────────────
#  Approval list (Phase 4 — /cycles/{id}/approvals)
# ─────────────────────────────────────────────────────────────────────────────
class ApprovalDetail(BaseModel):
    """One row in the approvals list — site summary + approval metadata."""
    id: int
    site: str
    status: str                        # ApprovalStatus value

    # Site aggregates (mirrors SiteSummary)
    staff: int
    payroll_current: float
    payroll_proposed: float
    hard_issues: int                   # fail-level compliance errors
    warn_count: int = 0               # warn-level issues not yet suppressed

    # Submission info
    submitted_by: str | None = None    # display name
    submitted_at: datetime | None = None
    submission_notes: str | None = None

    # Decision info
    decided_by: str | None = None
    decided_at: datetime | None = None
    decision_notes: str | None = None
